pkg/client/client_test.go

```diff
@@ -1910,7 +1946,303 @@ func TestClose(t *testing.T) {
+	s.NoError(err)
+	s.True(isTrackCalled)
+	s.Equal(1, len(s.mockProcessor.Events))
+	s.Equal("1212121", s.mockProcessor.Events[0].VisitorID)
+	s.Equal("15389410617", s.mockProcessor.Events[0].EventContext.ProjectID)
+}
+
+func (s *ClientTestSuiteTrackEvent) TestTrackWithNotificationAndEventTag() {
+
+	s.mockProcessor.On("ProcessEvent", mock.Anything).Return(true)
+	expectedUserContext := entities.UserContext{ID: "1212121", Attributes: map[string]interface{}{}}
+	expectedEvenTags := map[string]interface{}{
+		"client":  "ios",
+		"version": "7.0",
+	}
+	isTrackCalled := false
+	onTrack := func(eventKey string, userContext entities.UserContext, eventTags map[string]interface{}, conversionEvent event.ConversionEvent) {
+		isTrackCalled = true
+		s.Equal("sample_conversion", eventKey)
+		s.Equal(expectedUserContext, userContext)
+		s.Equal(expectedEvenTags, eventTags)
+		s.Equal(*s.mockProcessor.Events[0].Conversion, conversionEvent)
+	}
+
+	s.client.OnTrack(onTrack)
+	err := s.client.Track("sample_conversion", expectedUserContext, expectedEvenTags)
+
+	s.NoError(err)
+	s.True(isTrackCalled)
+	s.Equal(1, len(s.mockProcessor.Events))
+	s.Equal("1212121", s.mockProcessor.Events[0].VisitorID)
+	s.Equal("15389410617", s.mockProcessor.Events[0].EventContext.ProjectID)
+}
+
+func (s *ClientTestSuiteTrackEvent) TestTrackWithNotificationAndUserEvent() {
+
+	s.mockProcessor.On("ProcessEvent", mock.Anything).Return(true)
+	expectedUserContext := entities.UserContext{ID: "1212121", Attributes: map[string]interface{}{}}
+	expectedEventTags := map[string]interface{}{
+		"client":  "ios",
+		"version": "7.0",
+	}
+	isTrackCalled := false
+	onTrack := func(eventKey string, userContext entities.UserContext, eventTags map[string]interface{}, conversionEvent event.ConversionEvent) {
+		isTrackCalled = true
+		s.Equal("sample_conversion", eventKey)
+		s.Equal(expectedUserContext, userContext)
+		s.Equal(expectedEventTags, eventTags)
+		s.Equal(1, len(s.mockProcessor.Events))
+		s.Equal(*s.mockProcessor.Events[0].Conversion, conversionEvent)
+	}
+
+	s.client.OnTrack(onTrack)
+	err := s.client.Track("sample_conversion", expectedUserContext, expectedEventTags)
+
+	s.NoError(err)
+	s.True(isTrackCalled)
+	s.Equal(1, len(s.mockProcessor.Events))
+	s.Equal("1212121", s.mockProcessor.Events[0].VisitorID)
+	s.Equal("15389410617", s.mockProcessor.Events[0].EventContext.ProjectID)
+}
+
+func (s *ClientTestSuiteTrackEvent) TestTrackNotificationNotCalledWhenEventProcessorReturnsFalse() {
+	s.mockProcessor.On("ProcessEvent", mock.Anything).Return(false)
+
+	isTrackCalled := false
+	onTrack := func(eventKey string, userContext entities.UserContext, eventTags map[string]interface{}, conversionEvent event.ConversionEvent) {
+		isTrackCalled = true
+	}
+
+	s.client.OnTrack(onTrack)
+	err := s.client.Track("sample_conversion", entities.UserContext{ID: "1212121", Attributes: map[string]interface{}{}}, map[string]interface{}{})
+
+	s.NoError(err)
+	s.Equal(1, len(s.mockProcessor.Events))
```

Comment:
How it can be 1. I think this is confusing. The best way would be to mock event.CreateConversionEvent if possible. and return any invalid type and let ProcessEvent do its job. It will return error while invalid type is passed.